modcc/expression.cpp

```diff
@@ -690,18 +691,44 @@ void FunctionExpression::semantic(scope_type::symbol_map &global_symbols) {
         if(e->is_initial_block()) error("INITIAL block not allowed inside FUNCTION definition");
     }
 
-    // check that the last expression in the body was an assignment to
-    // the return placeholder
+    // check that the last expression in the body was an assignment expression
+    // or if/else expression that assigns the return placeholder
     bool last_expr_is_assign = false;
-    auto tail = body()->back()->is_assignment();
-    if(tail) {
+    if(auto tail = body()->back()->is_assignment()) {
         // we know that the tail is an assignment expression
         auto lhs = tail->lhs()->is_identifier();
         // use nullptr check followed by lazy name lookup
         if(lhs && lhs->name()==name()) {
             last_expr_is_assign = true;
         }
-    }
+    } else if(body()->back()->is_if()) {
+        last_expr_is_assign = true;
+        std::stack<expression_ptr> if_exp;
```

Comment:
I think that you can avoid the need for all of the clone and move below by using `std::stack<IfExpression*>`, which would also simplify later expressions...

```C++
else if (auto e = body()->back()->is_if()) {
  std::stack<IfExpression*> stack;
  stack.push(e);
  while (!stack.empty()) {
    auto tail = stack.top();
    stack.pop();
    if (auto branch = tail->true_branch()) {
        auto last = &branch->is_block->statements().back();
        if (last->is_assignment()) {
          // ...
        else if (last->is_if()) {
          stack.push(last->is_if());
        }
      // ...
    }
    ...
```

Reply:
Yes, that does look better. I'll change it.